pipeline/config.go

```diff
@@ -1,293 +1,49 @@
 package pipeline
 
 import (
-	"fmt"
-	"strings"
-
-	"github.com/observiq/stanza/errors"
 	"github.com/observiq/stanza/operator"
-	"github.com/observiq/stanza/operator/helper"
-	"github.com/observiq/stanza/plugin"
-	yaml "gopkg.in/yaml.v2"
 )
 
 // Config is the configuration of a pipeline.
-type Config []Params
-
-// BuildPipeline will build a pipeline from the config.
-func (c Config) BuildPipeline(context operator.BuildContext, pluginRegistry plugin.Registry, defaultOutput operator.Operator) (*DirectedPipeline, error) {
-	operatorConfigs, err := c.buildOperatorConfigs(pluginRegistry)
-	if err != nil {
-		return nil, err
-	}
-
-	operators, err := c.buildOperators(operatorConfigs, context)
-	if err != nil {
-		return nil, err
-	}
-
-	if defaultOutput != nil {
-		operators = append(operators, defaultOutput)
-	}
-
-	pipeline, err := NewDirectedPipeline(operators)
-	if err != nil {
-		return nil, err
-	}
-
-	return pipeline, nil
-}
-
-func (c Config) buildOperatorConfigs(pluginRegistry plugin.Registry) ([]operator.Config, error) {
-	operatorConfigs := make([]operator.Config, 0, len(c))
-
-	for i, params := range c {
-		if err := params.Validate(); err != nil {
-			return nil, errors.Wrap(err, "validate config params")
-		}
-
-		configs, err := params.BuildConfigs(pluginRegistry, "$", c.defaultOutput(i, "$"))
-		if err != nil {
-			return nil, errors.Wrap(err, "build operator configs")
-		}
-		operatorConfigs = append(operatorConfigs, configs...)
-	}
-
-	return operatorConfigs, nil
-}
-
-func (c Config) buildOperators(operatorConfigs []operator.Config, context operator.BuildContext) ([]operator.Operator, error) {
-	operators := make([]operator.Operator, 0, len(operatorConfigs))
-	for _, operatorConfig := range operatorConfigs {
-		operator, err := operatorConfig.Build(context)
+type Config []operator.Config
 
+func (c Config) BuildOperators(bc operator.BuildContext) ([]operator.Operator, error) {
+	operators := make([]operator.Operator, 0, len(c))
+	for i, builder := range c {
+		nbc := getBuildContextWithDefaultOutput(c, i, bc)
+		op, err := builder.BuildMulti(nbc)
```

Comment:
Seems like we're using `MultiBuilder` pretty much everywhere. Should we even expose the regular `Builder` struct anymore? If not, perhaps `MultiBuilder` should just be `Builder`, and the former `Builder` should be renamed to something internal.

Reply:
From the outside, yes, we are only using `MultiBuilder`. However, that's because we can make `Builder` implement `MultiBuilder` just by wrapping it. That way we don't have to change the interface of all the operator configs to return an array of operators rather than a single operator. 

I found it confusing that a `Build` method on a config would return multiple operators, that's why I made the `RegisterOperator` method take a `Builder` instead of a `MultiBuilder` like the `RegisterPlugin` method takes. 

It made sense to me, but if it just made it more confusing from an outside perspective, I can change it

Reply:
Seems fine to me. There's some nuance to it, but backwards compatibility is worth preserving when we can.